### DISCUSSION TITLE
Serve pocketbase from main function

### DISCUSSION THREAD

**Discussion by osadasami** [pocketbase/pocketbase#1345] [Q&A]
I have [Wails](https://wails.io/) app that looks like this

`main.go`
```go
func main() {
  app := NewApp()
  err := wails.Run(&options.App{})
)
```

Is it possible to serve pocketbase app from main function. Something like

```go
func main() {
  pocketbase.Serve("127.0.0.1:8090")
  app := NewApp()
  err := wails.Run(&options.App{}
)
```

I want to run wails app and pocketbase admin ui at the same time. It would allow me to make requests from wails frontend to pocketbase using javscript pocketbase sdk

From Wails frontend

```ts
export const pb = new PocketBase('http://127.0.0.1:8090')
```

I tried to use

```go
pocketBaseApp := pocketbase.New()

if err := pocketBaseApp.Start(); err != nil {
  log.Fatal(err)
}
```

but `127.0.0.1:8090` does not work

**Comment by osadasami** [comment]
Found the way

```shell
wails dev -appargs "serve"
```

But still interesting is it possible to serve pocketbase from main function?

**Reply by ganigeorgiev** [reply]
Yes, that could work. You can also optionally try to pass inside the string `--http` to change the default http serve address and port.

Option to customize the serve command default args programmatically is in the [roadmap](https://github.com/orgs/pocketbase/projects/2) and it will be added as part of the initialization config sometime in the future.

Side note: `pocketbase.Start()` starts an http server, aka. it is "blocking" so you may need to register it in a separate goroutine if you want to run it together with other code after `Start()`. aka.:

```go
func main() {
	go func() {
		app := pocketbase.New()
		
		if err := app.Start(); err != nil {
			log.Fatal(err)
		}
	}()
	
	// your other code (if it is not long running.blocking you may check sync.WaitGroup or other similar mechanism)...
}
```

**Reply by ganigeorgiev** [reply]
@osadasami ~~Another option could be also to trigger manually the `serve` command (_I haven't tested it_):~~

```go
cmd, _, _ := app.RootCmd.Find([]string{"serve"})
cmd.Execute()
```

**Update:** the above will not work since the app will be not bootstrapped yet, see below for an alternative.

**Reply by ganigeorgiev** [reply]
@osadasami Alternatively, you can also manually instantiate the `serve` command like this:

```go
app := pocketbase.New()
app.Bootstrap()
serveCmd := cmd.NewServeCommand(app, false)
serveCmd.Execute()
```

**Reply by osadasami** [reply]
Thank you so much! it works just perfectly

Now I can run dev server with pocketbase and even build wails app with pocketbase and when I run the app, pocketbase starts automatically

`main.go` looks like this

```go

func main() {
  app := NewApp()

  go func() {
    pocketBaseApp := pocketbase.New()
    pocketBaseApp.Bootstrap()
    serveCmd := cmd.NewServeCommand(pocketBaseApp, true)
    serveCmd.Execute()
  }()

  err := wails.Run(&options.App{
    Title:       "app-demo",
    Width:       1024,
    Height:      768,
    StartHidden: false,
    AssetServer: &assetserver.Options{
      Assets: assets,
    },
    BackgroundColour: &options.RGBA{R: 27, G: 38, B: 54, A: 1},
    OnStartup:        app.startup,
    Bind: []interface{}{
      app,
    },
  })

  if err != nil {
    println("Error:", err.Error())
  }
}
```

**Reply by oSethoum** [reply]
@osadasami this works but you still have to execute the executable from a terminal instance, but if you double click wails app to start the app just gonna exit immediately, right?.

**Reply by erasy1L** [reply]
> @osadasami this works but you still have to execute the executable from a terminal instance, but if you double click wails app to start the app just gonna exit immediately, right?.

Did you solve this?

**Reply by ganigeorgiev** [reply]
@erazr You are commenting to a very old discussion and a lot of things has changed since then, including manually starting the PocketBase server, see the [`apis.Serve`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.22.18/apis#Serve) method (_or the comment below_).

If you have other questions, please open a new Q&A discussion elaborating more on your use case and what you are trying to do.

**Comment by avr-main** [comment]
I got it working with this

in the **main.go** file
add this to imports

    import (
	    "embed"
    
	    "github.com/wailsapp/wails/v2"
	    "github.com/wailsapp/wails/v2/pkg/options"
	    "github.com/wailsapp/wails/v2/pkg/options/assetserver"
    
	    //POCKETBASE
	    "github.com/pocketbase/pocketbase"
	    "github.com/pocketbase/pocketbase/apis"
    )
in the **go func()**

    go  func() {
	    pocketBaseApp  :=  pocketbase.New()
	    pocketBaseApp.Bootstrap()
	    apis.Serve(pocketBaseApp, apis.ServeConfig{
		    HttpAddr: "127.0.0.1:8080",
		    ShowStartBanner: false,
		})
	}()
if you want to access the backend pocketbase adminUI,  
visit  
**http://wails.localhost:8080/_/**

**Reply by PylotLight** [reply]
Testing this in v3 has not been successful for me, do you have a repo I could look at for this?
Adding above code doesn't seem to be reachable on the wails endpoint so far.

**Comment by yharby** [comment]
any updates from anyone managed to do that here ? thanks !

**Comment by avr-main** [comment]
@PylotLight I am so sorry for the Extremly late response
@yharby Here you go

[Repo](https://github.com/avr-main/Wails-PocketBase-Template)

Added details in README and comments
